[Compose 페이먼츠 4단계] 이든 미션 제출합니다.

devfeijoa opened the pull request:
# 셀프 체크리스트

 <!--
 모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
 -->

- [x] 프로그램이 정상적으로 작동하는가?
- [ ] 모든 테스트가 통과하는가?
- [x] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

안녕하세요 페토 ! step4 리뷰도 잘 부탁드립니다 !

 <!--
 학습 목표에 맞게 구현하였는지 점검하고, 피드백을 받을 준비가 되었는지 확인하세요.
 정답을 묻기보다, 자신의 생각과 의사 결정 과정을 공유하며 피드백을 통해 논의하는 것이 좋습니다.
 -->
 
-

## 코드 리뷰 커뮤니케이션

 <!--
 리뷰어에게 질문이 있다면, PR에 직접 코멘트를 남겨 보세요.  
 특정 코드 줄에 의견을 남기면 보다 구체적인 피드백을 받을 수 있습니다.  
 -->

📌 [GitHub에서 PR에 댓글 남기는 방법](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request)

## 참고 자료

 <!-- 관련 문서나 참고할 만한 자료를 추가하세요. -->

## 스크린샷

 <!-- 필요하다면 스크린샷을 첨부하세요. -->


https://github.com/user-attachments/assets/7919f7d2-bf68-444b-88a4-d15a6e910820


![]()

devfeijoa committed:
docs: step4 카드 수정 기능 요구 사항 기능 명세서 작성

devfeijoa committed:
refactor: `NewCardTopBar` 컴포저블 title 파라미터 추가 및 프리뷰 업데이트

devfeijoa committed:
feat: 카드 수정 기능 추가

devfeijoa committed:
refactor: 카드 추가 및 수정 로직 개선

devfeijoa committed:
feat: 카드 수정 기능 추가

devfeijoa committed:
feat: 카드 수정 기능 추가

- `NewCardScreenActivity`에서 기존 카드 정보를 받아올 수 있도록 수정했습니다.
- 카드 정보를 수정하고 저장하면, `CardWalletScreen`에서 해당 카드의 정보가 업데이트됩니다.
- `NewCardScreenActivity`의 `newIntent` 함수에 `existingCard`와 `index` 파라미터를 추가하여, 카드 수정 시 기존 카드 정보와 인덱스를 전달할 수 있도록 했습니다.

devfeijoa committed:
feat: 등록된 카드 클릭 이벤트 핸들러 추가

devfeijoa committed:
feat: 카드 목록 화면에서 카드 선택 시 해당 카드 정보로 카드 추가 화면 이동

- `CardWalletScreen`에서 등록된 카드를 클릭하면 `NewCardScreenActivity`로 이동하도록 구현했습니다.
- 이동 시 선택된 카드의 정보(`existingCard`)와 목록에서의 인덱스(`index`)를 전달합니다.
- 카드가 하나만 있을 때도 카드 클릭 시 수정 화면으로 이동하도록 수정했습니다.

devfeijoa committed:
docs: 기능명세서 업데이트

chanho0908 commented on app/src/main/java/woowacourse/payments/ui/newcard/NewCardScreen.kt:
```diff
@@ -45,10 +50,26 @@ fun NewCardScreen(
+                    it.cardHolderName != cardHolderName ||
+                    it.cardPassword != cardPassword ||
+                    it.cardCompanyUiModel != selectedCardCompany
+            } == true
```
컴포저블은 화면을 그리는 책임만 할 수 있도록 책임 여부에 대한 판별은 CardUiModel의 책임으로 주는건 어떨까요 ?

devfeijoa replied:
기존 카드 정보 비교 로직의 책임을 CardUiModel로 이동하였습니다 ! aeaaec94043e237d55716ca06237ad01520032c2

chanho0908 commented on app/src/main/java/woowacourse/payments/ui/cardwallet/CardWalletScreen.kt:
```diff
@@ -115,8 +113,19 @@ fun CardWalletScreen(
+                    repeat(cardList.size) { index ->
+                        RegisteredCard(
+                            cardUiModel = cardList[index],
+                            onClick = {
+                                val intent =
+                                    NewCardScreenActivity.newIntent(
+                                        context,
+                                        existingCard = cardList[index],
+                                        index = index,
+                                    )
+                                cardAddLauncher.launch(intent)
+                            },
```
강의 시간에도 코치님들이 이야기 했던 것 같은데 화면 이동에 대한 책임은 Activity의 역할일까요 ? 컴포저블의 역할일까요 ?

devfeijoa replied:
화면 이동에 대한 책임은 Activity에 대한 책임이라는 의도실까요..? 😢

chanho0908 replied:
이든의 생각을 여쭤본거에요 !

chanho0908 reviewed (changes requested):
안녕하세요 이든 ! 4단계도 고생 많으셨습니다 👍 💯 

아직 미완성인 부분이 있는 것 같아 해당 문제들 먼저 해결한뒤 리뷰 부탁드려요 😃 
+ 카드가 한 장일 때 카드 수정 화면으로 이동하지 않습니다.
+ 카드가 두 장일 때 수정 화면으로 이동하지만 수정할 카드의 정보가 넘어오지 않아요 !
+ 카드 목록 화면에서 화면이 회전되면 모든 카드가 사라집니다.
+ 카드 추가 화면에서 카드사를 선택하고 화면을 회전하면 카드를 선택하는 바텀시트가 또 올려옵니다

devfeijoa committed:
fix: 카드가 한 장 일때도 수정 화면으로 이동하도록 수정

devfeijoa committed:
refactor: `rememberSaveable` 적용하여 카드 수정 로직 개선

- `NewCardScreen`의 입력 필드 상태를 `rememberSaveable`로 변경하여 화면 회전 시에도 상태를 유지하도록 했습니다.

- 카드 수정 모드일 때, 초기에는 카드사 선택 BottomSheet가 표시되지 않도록 수정했습니다.

- 수정할 카드의 정보가 넘어오지 않는 문제 수정

devfeijoa committed:
refactor: 기존 카드 정보 비교 로직의 책임을 CardUiModel로 이동

- `NewCardScreen`에서 기존 카드 정보와 현재 입력된 정보를 비교하는 로직을 `isDifferentFrom` 함수를 사용하도록 변경했습니다.

devfeijoa committed:
refactor: 카드 편집 화면에서 카드사 선택 바텀 시트가 불필요하게 표시되는 문제 수정

- 기존 카드가 있는 경우 카드사 선택 바텀 시트가 표시되지 않도록 수정했습니다.

devfeijoa committed:
refactor: `rememberSaveable` 적용하여 화면 회전시에 초기화되지 않도록 수정

devfeijoa commented:
해당 리뷰 다 반영하여 수정하였습니다 !

1. 카드가 한 장일 때 카드 수정 화면으로 이동하지 않습니다. -> bceb564e4515818ab506e38880fa796fbfc55e87 
2. 카드가 두 장일 때 수정 화면으로 이동하지만 수정할 카드의 정보가 넘어오지 않아요 ! -> 864f03a41233c648cb5cb236c9256a316c19f03a
3. 카드 목록 화면에서 화면이 회전되면 모든 카드가 사라집니다. -> 93498548a88e9327fdd8dbe7c47d1c3cd719f38c
4. 카드 추가 화면에서 카드사를 선택하고 화면을 회전하면 카드를 선택하는 바텀시트가 또 올려옵니다 -> 96a9ecd4367ad954d93778dcc4d1d0ab15b3562d

chanho0908 commented on app/src/main/java/woowacourse/payments/ui/component/NewCardTopBar.kt:
```diff
@@ -17,12 +17,14 @@ import woowacourse.payments.R
     onBackClick: () -> Unit,
     onSaveClick: () -> Unit,
-    modifier: Modifier = Modifier,
+    isSaveEnabled: Boolean = true,
```
`isSaveEnabled`가 저장 여부를 나타내는 인자라면 기본값을 false로 두는게 자연스럽지 않을까요 ?

카드를 수정하던, 추가하던 특정 조건이 만족되어야 저장이 가능하니까요 !

devfeijoa replied:
해당 인자 필요하지 않게 되어 삭제하였습니다 !

chanho0908 commented on app/src/main/java/woowacourse/payments/ui/newcard/NewCardScreen.kt:
```diff
@@ -30,25 +31,46 @@ import woowacourse.payments.ui.model.CardUiModel
-    var cardExpiryDate by remember { mutableStateOf("") }
-    var cardHolderName by remember { mutableStateOf("") }
-    var cardPassword by remember { mutableStateOf("") }
+    val isEditMode = existingCard != null
```
이든께선 수정/추가 화면을 하나의 컴포저블로 재활용해주셨군요 !
강의시간에 언급되었던 내용인 만큼 이든께서 재활용하신 이유가 궁금해요 😃

devfeijoa replied:
`카드 추가` 화면과 `카드 수정` 화면이 거의 같은 화면 이기 때문에 중복된 UI를 두 번 만들 필요가 없다고 생각했습니다.
따라서 existingCard 여부로 분기 처리 했습니다 !

chanho0908 commented on app/src/main/java/woowacourse/payments/ui/newcard/NewCardScreen.kt:
```diff
@@ -30,25 +31,46 @@ import woowacourse.payments.ui.model.CardUiModel
+    var showBottomSheet by rememberSaveable { mutableStateOf(false) }
 
     LaunchedEffect(Unit) {
-        showBottomSheet = true
+        if (existingCard == null) {
+            showBottomSheet = true
+        }
     }
```
`LaunchedEffect`의 `key`는 어떤 역할을 하나요 ?

devfeijoa replied:
`LaunchedEffect`는 한 번만 실행할 비동기 블록을 정의할 때 썼고, `key`를 기준으로 언제 다시 실행할 지 결정합니다.
따라서`key`가 바뀔 때 마다 내부 코루틴이 cancel + restart 됩니다.

여기서는 화면이 처음 그려질 때 bottomSheet를 열어라는 의미 였습니다 !

chanho0908 commented on app/src/main/java/woowacourse/payments/ui/newcard/NewCardScreen.kt:
```diff
@@ -61,6 +83,7 @@ fun NewCardScreen(
+                isSaveEnabled = isChanges,
             )
         },
     ) { innerPadding ->
```
`Scaffold`의 하위 컴포저블에 `innerPadding`을 전달하지 않으면 컴파일러가 오류를 발생시킬텐데
이 이유에 대해서 알고 계실까요 ?

devfeijoa replied:
`Scaffold`는 내부적은로 Appbar, NavigationBar 와의 레이아웃 충돌을 막기 위해 자식 컴포저블에 padding 정보를 내려줍니다.
만약 innerPadding을 생략하면 Compose는 이 padding 값을 쓰고 있지 않다고 경고하거나 경우에 따라 빌드 에러로 처리한다고 합니다.

페토가 더 알고 있는 부분이나 제가 잘못 알고있는 부분이 있다면 알려주시면 감사하겠습니다 👍

chanho0908 commented on app/src/main/java/woowacourse/payments/ui/newcard/NewCardScreen.kt:
```diff
@@ -30,25 +31,46 @@ import woowacourse.payments.ui.model.CardUiModel
-    var cardPassword by remember { mutableStateOf("") }
+    val isEditMode = existingCard != null
+
+    var cardNumber by rememberSaveable { mutableStateOf(existingCard?.cardNumber ?: "") }
+    var cardExpiryDate by rememberSaveable { mutableStateOf(existingCard?.cardExpiryDate ?: "") }
+    var cardHolderName by rememberSaveable { mutableStateOf(existingCard?.cardHolderName ?: "") }
+    var cardPassword by rememberSaveable { mutableStateOf(existingCard?.cardPassword ?: "") }
 
     var selectedCardCompany by rememberSaveable { mutableStateOf<CardCompanyUiModel?>(null) }
```
이제 이든께서 각각의 상태 하나씩 관리하는 방법에 익숙해지신 것 같으니 이제 이 아이들을 하나의 객체로 관리하는걸 연습해보면 좋을 것 같아요 👍

devfeijoa replied:
하나의 객체로 관리하도록 수정하였습니다 ! 9125149c2656d05b24d3010fc39618bbb0974ec6

chanho0908 reviewed (changes requested):
안녕하세요 이든 ! 리뷰를 빠르게 적용해주셨네요 👍 

수정할만한 부분들을 남겨놨으니 편하게 적용해주세요 !

devfeijoa committed:
fix: 카드 수정 시 변경 사항이 있을 때만 저장 버튼 활성화

- 카드 수정 화면(`NewCardScreen`)에서 기존 카드 정보와 변경 사항이 있을 경우에만 '저장' 버튼이 활성화되도록 로직을 수정했습니다.
- 카드 신규 등록 시에는 이 로직이 적용되지 않습니다.

devfeijoa committed:
fix: NewCardScreen에서 isSaveEnabled 속성 제거

`TopAppBar`의 `isSaveEnabled` 속성을 제거하여 불필요한 저장을 방지합니다.

devfeijoa committed:
refactor: NewCardScreen 상태 관리 로직 리팩터링

- 여러 `mutableStateOf`로 관리되던 카드 정보 관련 상태들을 `CardUiModel` 데이터 클래스를 사용하여 단일 상태로 통합했습니다.
- 각 입력 필드의 `onValueChange` 콜백에서 `copy`를 사용하여 상태를 업데이트하도록 수정했습니다.

devfeijoa committed:
feat: 카드 수정 시 카드 정보가 표시되도록 구현

- 카드 수정 모드(`isEditMode`)일 경우, `RegisteredCard` 컴포저블을 사용하여 카드 번호, 소유자 이름, 만료일이 포함된 카드 UI를 표시하도록 `NewCardScreen`을 수정했습니다.
- `PaymentCard` 컴포저블의 파라미터를 `cardUiModel`로 변경하여 카드 정보를 받도록 수정했습니다.

devfeijoa committed:
refactor: NewCardTopBar에서 isSaveEnabled 파라미터 제거

devfeijoa committed:
style: 코드 포맷팅 및 불필요한 import 제거

- 람다 파라미터 `it`을 명시적으로 추가했습니다.
- 불필요한 import 문을 제거했습니다.
- 코드 포맷팅을 적용했습니다.

devfeijoa commented:
변경 사항이 없으면 수정되지 않아야 하는 요구사항도 반영하여 수정하였습니다 ! 페이먼츠 미션동안 페토의 소중한 시간내어 리뷰해주셔서 감사합니다 👍

chanho0908 reviewed (approved):
컴포즈 미션동안 고생하셨습니다 이든 ☺️
이든덕에 리뷰어로써 어떻게 리뷰를 해야할지 많이 배웠어요 ! 

다음 미션도 응원할게요 파이팅 ! 🙂
